### ISSUE TITLE
HDR AVIF to JXR conversion

### ISSUE THREAD

**Issue by nbryant42** [ImageMagick/ImageMagick#6642]
### Is your feature request related to a problem? Please describe.

Currently, conversion of HDR AVIF to JXR fails to retain correct luminosity.

To reproduce:

1. Export an HDR AVIF with the Technology Preview mode in recent versions of Adobe Camera Raw (I can provide samples of my own photos if necessary). These can currently be viewed correctly with Google Chrome and edited with Krita (with some issues on re-save perhaps)
1. Attempt to convert, e.g. `magick filename.avif filename.jxr`
1. Observe that the resulting JXR does not render correctly in Microsoft Photos (as of version 2023.11050.16005.0) despite Photos having support for JXR viewing.

### Describe the solution you'd like

`magick filename.avif filename.jxr` should produce usable JXR's, ideally with no non-default command line options

**Comment by urban-warrior** [comment]
ImageMagick employs a delegate program called `JxrEncApp` for JXR writing. To anticipate advancements in addressing your concerns, notify the libjxr developers. ImageMagick will automatically gain from any enhancements implemented by the libjxr developers.

**Comment by nbryant42** [comment]
@urban-warrior, I am aware of that, but I have no way of knowing whether the problem lies in ImageMagick's input to `JxrEncApp` or within `JxrEncApp` itself.

**Comment by urban-warrior** [comment]
ImageMagick converts any image to BMP and passes it to JxrEncApp.   You can edit `delegates.xml` and instead use TIFF to get better fidelity as TIFF supports the full 10-bits of AVIF, whereas BMP is 8-bit.  Any other issues lies within the `JxrEncApp` utility and must be addressed by the libjxr developers.

**Comment by nbryant42** [comment]
Thanks for the hint. I get pretty good fidelity converting HDR AVIF-->TIFF using Krita. Simply open the AVIF as saved by Camera Raw and Save As TIFF. The resulting TIFF will open back up in Krita and display correctly as HDR. Krita opens these TIFF's in its "Krita Display P3 Linear" color space as 32bpc float.

`magick filename.avif filename.tif` doesn't do so well by default. It appears to clip everything down to an SDR sRGB colorspace at 16bpc. `magick -colorspace DisplayP3` is no better as the resulting TIFF still opens as sRGB in Krita. The `-storage-type float` option documented at https://imagemagick.org/script/command-line-options.php doesn't appear to work, at least not for TIFF output. I'd assume that 32bpc float (allowing values >1.0 for luma brighter than SDR white) would be the simplest option for transfer into programs like `JxrEncApp`.

So far I'm not sure how to get `JxrEncApp` to deal with the Krita-produced TIFF's. It just writes an empty JXR. So there seem to be problems throughout the toolchain.

**Comment by fmw42** [comment]
Try adding defines for float format tiff.  See -define quantum:format=floating-point

**Comment by urban-warrior** [comment]
Use this command as @fmw42 suggests:

```
$ magick triad.avif -define quantum:format=floating-point triad.tif
$ compare -metric rmse triad.tif triad.avif null:
0.347461 (5.30191e-06)
```

Notice near non-existant distortion.

For JXR, if you change the `delegates.xml` file to use TIFF as the intermediate file instead of BMP, we get minimal distortion as well.  We'll add a patch to to fix this issue later today:

```
$ magick triad.avif -define quantum:format=floating-point triad.jxr
$ compare -metric rmse triad.jxr triad.avif null:
0.347461 (5.30191e-06)
```

**Comment by nbryant42** [comment]
Thanks. I built the latest source including your commit. This helps: now I'm able to get JXRs that are at least not corrupted. They're readable by Microsoft Photos, but it looks like the luma has been compressed into the SDR range. Based on your comments above about the `compare` results, ImageMagick thinks the file match, thus my suspicion is that this has to do with the way ImageMagick *reads* the input AVIF into its internal data model.

Downloading `Triad-hdr.avif` from the Greg Benz site, and using `magick Triad-hdr.avif -define quantum:format=floating-point Triad-hdr.tif` confirms that ImageMagic writes out TIF's that don't render luminosity in the same brightness range as the original AVIF. Confirmed this by opening the original AVIF and converted TIF in both Krita and Camera Raw on Windows with an HDR display; input and output file don't render the same way; the output. To my eye, it looks like the pixel data is not *clipped*, but rather *scaled down* into the SDR range.

I'm not an expert on HDR file formats but from what I've been able to glean, there are at least two things that need to be accounted for: color space (should be P3 but the output TIF loads as sRGB in Krita) and transfer function (but we're using 16-bit float, which is linear, so the transfer function shouldn't matter, right?)

**Comment by urban-warrior** [comment]
We use libheif to read AVIF images.  The pixel values are defined by whatever that delegate library returns.  We also include any color profile included in the source image.  You can use the `-profile` option to transform to an alternate colorspace such as P3.  The Triad image does not appear to include a source color profile.  It does include an EXIF and XMP profile.

**Comment by nbryant42** [comment]
Understood. However, I have some of my own photography which was processed from Nikon RAW (NIF) to AVIF via Camera Raw in a P3 colorspace, but when using `magick filename.avif -define quantum:format=floating-point filename.tif` the TIF loads into Krita as sRGB. Seems nearly certain this is incorrect. The brightness is also too dim overall, same as for the Triad example.

**Comment by urban-warrior** [comment]
We have less control over the workflow when we leverage delegate libraries and programs.  libheif is returning AVIF pixels in the sRGB colorspace.  You can transform the image to [P3](http://color.support/iccprofiles.html) after you read it.  TIFF should capture the P3 profile as part of its metadata and perhaps Krita will be responsive to the profile.

**Comment by nbryant42** [comment]
Yes, my experience has been that when TIFF's are written directly by Camera Raw or Krita, these programs will set the right profile.

**Comment by nbryant42** [comment]
`coders/heic.c` reads an unsigned short value via `ScaleShortToQuantum()`, which, when `MAGICKCORE_HDRI_SUPPORT` is defined, simply floating-point divides by 257.0, so it can never return a value greater than 255. I'm really a newbie to HDR data formats but I thought the idea was to represent anything brighter than the SDR whitepoint by using values greater than 255 (int) or 1.0 (float); either (1) my understanding is not correct or (2) ImageMagick's data model isn't up to the task or (3) the data model is right but there is some issue with metadata tagging on output?

**Comment by urban-warrior** [comment]
See https://imagemagick.org/script/high-dynamic-range.php.  Make sure you have the default build of ImageMagick which is Q16-HDRI.  Libheic returns 8, 10, or 12 bit pixels.  Your AVIF image is 10-bit so we multiply to expand to 16-bit internally rather than divide.  Its possible the utility that created the image used a clever profile to map out of band pixels to in-band such as P3.  If so you need to set the profile to P3, `-set profile myP3.icc`, and/or transform it to some other colorspace if your workflow requires it, `-profile sRGB.icc`.

**Comment by nbryant42** [comment]
I'm building ImageMagick github main branch on Ubuntu 22.04 with Ubuntu's version of libheif, which is 1.12.0 as of that version of Ubuntu. Configured with `./configure --enable-hdri`. libtiff-4 also installed via Ubuntu default repo.

I read through the libheif changelog on Github and there are some newer versions available but no changes that seem relevant to this issue.

You might want to have a look at `coders/heic.c` because it only has two code paths on decode. Either libheif returns 8bpc, in which case it calls `ScaleCharToQuantum`. Otherwise you get this, and ScaleShortToQuantum definitely divides by 257.

```c
      /*
        Transform 10-bit or 12-bit image.
      */
      q=QueueAuthenticPixels(image,0,y,image->columns,1,exception);
      if (q == (Quantum *) NULL)
        break;
      p=pixels+(y*stride);
      for (x=0; x < (ssize_t) image->columns; x++)
      {
        unsigned short pixel = (((unsigned short) *(p+1) << 8) |
          (*(p+0))) << shift; p+=2;
        SetPixelRed(image,ScaleShortToQuantum(pixel),q);
```

Maybe there's a multiplication later in your processing flow but just looking at this, it's a division.

**Comment by snibgo** [comment]
>  I'm really a newbie to HDR data formats but I thought the idea was to represent anything brighter than the SDR whitepoint by using values greater than 255 (int) or 1.0 (float);

That is roughly true, but omits important detail.

In particular, when an 8-bit integer format is used, there are no values greater then 255. Similarly 10-bit (1023) and 12-bit (4095).

In heic.c function ReadHEICImageHandle(), IM maps values from 0 to 255 (or 1023 or 2095) to 0 to QuantumRange (which is 65535 for Q16). This is a multiply. Note that the C operation "<<" is a bit-shift left, which multiplies.

Some metadata in the file will specify the transfer function (ie how to transform the pixel values into linear values). It might also define the HDR overhead assumed by the image author, eg 3 stops.

> libheif is returning AVIF pixels in the sRGB colorspace.

If so, then the transfer function is also sRGB, I suppose.

If your IM is HDRI, a reasonable action after reading the AVIF is to convert to linear, and multiply up so the SDR threshold is 100% or QuantumRange, eg 3 stops would need multiplying by 8.

**Comment by urban-warrior** [comment]
Where's the divide?  10-bit image pixels consume two bytes.  We take the first byte and shift by 8 and or in the second byte shifted by (16-10).  This expands the 10-bit range to 16-bits.  ScaleShortToQuantum() does not scale the short to 0..255.  It is essentially a no-op.  Look @ `MagickCore/quantum-private.h` for the `MAGICKCORE_QUANTUM_DEPTH == 16` case.  You'll see that ScaleShortToQuantum() is defined as

```
static inline Quantum ScaleShortToQuantum(const unsigned short value)
{
  return((Quantum) value);
}
```

Note, no divide.

**Comment by nbryant42** [comment]
OK. I'll take everything you both just said on faith for now, maybe the real problem is some TIFtags that aren't being written

**Comment by kmilos** [comment]
> The Triad image does not appear to include a source color profile.

It does.

AVIF/HEIF and JPEG XL can (and will even prefer) to encode the color their color profile in a more compact [CICP](https://en.wikipedia.org/wiki/Coding-independent_code_points) representation rather than ICC. In the case of AVIF/HEF, this is stored in the `nclx` box, and libheif will happily extract this for you. I suspect IM does not use this information for now?

For the `Traid-hdr.avif` file, this CICP information is as follows (from exiftool):
```
ColorProfiles                   : nclx
ColorPrimaries                  : SMPTE EG 432-1
TransferCharacteristics         : SMPTE ST 2084, ITU BT.2100 PQ
MatrixCoefficients              : BT.601
```
Which means you cannot assign just any "P3" profile to this, it has to be the one using a PQ transfer curve.

**Comment by ACIXCIX0001** [comment]
how to convert .jxr to a workable and editable format by keeping the hdr 16 bpc rec. 2100 color space st. 2084 pq tranfert function ?  
  
.jxr hdr screenshots are generated by [xbox gamebar, nvidia shadowplay, obs]  
tried everything :  
[jxrlib](https://github.com/Knewest/precompiled-jxrlib-binaries)  
`jxrdecapp.exe -i 'obs.jxr' -o 'obs_jxrdecappconvert.tif'`  
open and exported to [.tif 16, .exr 16, .png 16] with [photoshop plugin](https://www.microsoft.com/en-us/download/details.aspx?id=52369)  
open and exported to [.tif 16, .exr 16, .png 16] with [affinity photo 2]  
  
asked almost every community forums [colorists, video editing, photo editing, .jxl, jxrlib binaries compiler, .......], no one appears to have a clue on how to process hdr .jxr images out and in of their original container.  
because indeed ! as soon as the image is [extracted, converted] out of it s original container, admittedly it keeps it s 16 bpc bit depth but it absolutely does not keep the original [rec. 2100 color space, st. 2084 pq tranfert function] metadata !  
and then, it s done ! the image has in all cases been converted to [srgb color space, gamma 2.2 transfert function], even if you would try to mark the resulting [.tif, .exr, .hdr, .png] metadata with the original hdr [rec. 2100 color space, st. 2084 pq tranfert function] metadata, the information is loast and the image will still appears washed out and dull no matter the viewer used \[['microsoft photos'](https://apps.microsoft.com/detail/9WZDNCRFJBH4?hl=en-US&gl=US), ['hdr + wcg image viewer'](https://apps.microsoft.com/detail/9PGN3NWPBWL9?hl=en-US&gl=US)\]  
  
here are some sample and you can notice all the original .jxr displays correctly, but none of the converted to other formats does :  
[onedrive](https://1drv.ms/f/s!AiUmCeW6ezhq90iZ1PpOJbwDiVp6?e=j6DCR9)

**Comment by AngelaDMerkel** [comment]
I have brought up a similar issue with respect to HDR JXL encoding and it seems like `libultrahdr` will offer a partial fix for this, but my understanding is that AVIF and JPEG both rely on gain maps instead of being encoded directly in HDR. Exporting a HDR JXL image and then converting to JXR might work better but support for manipulating HDR images in either of it's iterations is likely a long way off. Google and Adobe seem like they are harmonising their approaches, but who knows how long until we reach that point.

**Comment by nbryant42** [comment]
I believe that's not 100% correct. AVIF has a native HDR10 mode, no gain maps. I don't even know whether it supports gain maps in addition. But in the case of JPEG I think you're right, it's only done via gain maps.

The problem with gain maps is that at the moment, there are some competing standards on iOS and Android, and no broad support for the Adobe standard. See https://gregbenzphotography.com/hdr-photos/instagram-now-supports-hdr-photos/ for a bit more discussion on this point.

**Comment by AngelaDMerkel** [comment]
> AVIF has a native HDR10 mode, with gain maps being optional (and not yet broadly supported)

You are correct, but a talk given my Eric Chan who leads the Adobe development of Lightroom and the HDR tools indicated that they plan to move fully over to gain maps. As it stands you're correct though. I would like if there were a more universal standard, but alas.
